apps/omg/lib/omg/state/core.ex

```diff
@@ -161,22 +183,36 @@ defmodule OMG.State.Core do
   """
   @spec exec(state :: t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
           {:ok, {Transaction.tx_hash(), pos_integer, non_neg_integer}, t()}
-          | {{:error, Validator.exec_error()}, t()}
+          | {{:error, Validator.can_process_tx_error()}, t()}
   def exec(%Core{} = state, %Transaction.Recovered{} = tx, fees) do
     tx_hash = Transaction.raw_txhash(tx)
 
-    case Validator.can_apply_spend(state, tx, fees) do
-      true ->
-        {:ok, {tx_hash, state.height, state.tx_index},
-         state
-         |> apply_spend(tx)
-         |> add_pending_tx(tx)}
-
+    with {:ok, fees_paid} <- Validator.can_process_tx(state, tx, fees) do
```

Comment:
> That's why case's exist.

No, this  is not the case :joy: 
IMO, `case` vs `with` is often misunderstood and the article you mention is very basic in its conclusions. You can probably use `case` and `with` interchangeably, but they are different at the __intention__ layer.

`case` is general-purpose branching concept it shows several equally important cases where flow can branch.

`with` is a concept of railway oriented programming pattern so there is happy-path flow and the error flow. In the happy-path functions use the results of the previous executions when successful to produce the final result. The first error is returned otherwise.

There is a great work of Scott Wlaschin who describes many of FP concepts (however in F# lang)

Reply:
I would disagree, but I'll let it pass since most of this code will get removed or rewritten.